Global Styles: Add Style Book to Global Styles

noisysocks opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Resolves https://github.com/WordPress/gutenberg/issues/44420.

Adds a new Style Book feature to Global Styles in the Site Editor. This displays a preview of all block types in a resizable editor frame.

## Why?
From https://github.com/WordPress/gutenberg/issues/44420:

> It's not uncommon to find yourself styling a block in Global Styles without that block being present on the canvas. Consequently you're left unsure as to whether the desired outcome is achieved.
>
> One solution to this could be providing a dedicated document containing all blocks, that users can observe on demand. 

## How?
A new component called `StyleBook` is added to `@wordpress/edit-site`. This uses the existing `BlockPreview` component to display a preview of every registered block in an iframe per block.

Each iframe has the site styles loaded into it, allowing you to preview them. The background colour of `StyleBook` is matched with the theme using `useStyle( 'color.background' )`. This all works because `StyleBook` is a descendent of `BlockEditorSettings` and `GlobalStylesProvider`.

Clicking on a block preview opens that block in Global Styles. This is done by having `StyleBook` be a `Fill` rendered by `GlobalStylesUI`. That lets us have access to the Global Styles navigation state using `useNavigator`. The style book is then mounted into the correct part of the editor DOM using `StyleBook.Slot`.

## Testing Instructions
1. Go to Appearance → Editor.
2. Open the (Global) Styles panel.
3. Click on the new _Open Style Book_ button. (Illuminati eye icon.)
4. Have fun.

## Screenshots or screencast 

https://user-images.githubusercontent.com/612155/204726373-23d31c97-1ab7-4678-a6db-7799a2839f7f.mp4

ramonjd commented on packages/edit-site/src/components/style-book/index.js:
```diff
@@ -0,0 +1,88 @@
+						} }
+					>
+						<div className="edit-site-style-book__blocks">
+							{ getBlockTypes()
```
Nice. I wonder if there's a way to memoize this based on the activeCategory (`tab.name`) I tried to memoize the `.edit-site-style-book__blocks` element, but it's a bit tough as it's re-rendered inside the tab regardless (and makes it worse 🤣 )

noisysocks replied:
Yeah could do. It might make switching to a tab _the second time_ a little bit faster. I'll have a play with performance after we land on a final design.

noisysocks replied:
I suspect most of the sluggishness comes from having so many frames and resize listeners 😀 I don't think there's a lot we can do at the React component level to fix that. I did notice that when I add `transition: background .3s` to the canvas that things _feel_ less sluggish, especially in Chrome. It think it's good enough for a v1.

ramonjd replied:
Looks good, and agree it's a good candidate for v1. ❤️ 

Should we place it behind an experimental flag?

I wonder if we can get around the iframe issue by rendering the block preview using the [useBlockPreview hook](https://github.com/WordPress/gutenberg/blob/abc8d7e850e0e6162725fe22f1b1d033f3d22ee4/packages/block-editor/src/components/block-preview/index.js#L82-L81) 

I tried it out, basically replicating [this unit test](https://github.com/WordPress/gutenberg/blob/d50e613d110ea7ccd5de3f647c8dcf3b72bdda74/packages/block-editor/src/components/block-preview/test/index.js#L84) and 🤷 

https://user-images.githubusercontent.com/6458278/203697141-16b31f9d-875f-44c7-80f6-4557bdf67b2d.mp4

No iframes at least, but not sure about implications re: styles 😄

noisysocks replied:
The iframes are what the styles are loaded into, hence why they appear unstyled for you there. Potentially we could make all of the previews go into a single iframe which should be less intensive on the browser. Can do that as a follow-up if necessary.

ramonjd replied:
> The iframes are what the styles are loaded into, hence why they appear unstyled for you there.  Potentially we could make all of the previews go into a single iframe which should be less intensive on the browser

Ah okay, thanks for the explainer. Makes sense.

noisysocks commented:
This is taking shape now! I updated the video in the description and listed my remaining todos which I'll work on next week.

javierarce commented:
I'm having some problems testing this PR due to some storybook errors (here's [my log](https://gist.github.com/javierarce/9a990df52c0bbde43a766391bd072fa8) after running `npm run dev`)

javierarce commented:
One thing I noticed in the video is the placement of the ellipsis menu and the eye icon: we should show the eye first and then the ellipsis.

noisysocks commented:
Oops sorry @javierarce I'll fix that up.

noisysocks commented:
I stacked this onto https://github.com/WordPress/gutenberg/pull/46099. @javierarce you should be able to build the branch with no compile errors now.

javierarce commented:
Thanks!

Ok, this looks great, but I found some things:

- I think we should change the pointer style to the hand when hovering a block
- If you click several blocks in the canvas and then use the Global Styles sidebar back button to reach the top level (or the list of blocks), you'll be presented with every single block settings panel you clicked before. I think we shouldn't stack those clicks. 
- We should highlight the block on the canvas if you select it in the sidebar.
- The content within each section should align vertically with the label:
<img width="957" alt="Screenshot 2022-11-28 at 11 49 18" src="https://user-images.githubusercontent.com/4933/204259395-91e9955b-0ab4-4ffd-a469-ca03f889f043.png">

jasmussen commented:
Thanks all for working on this, and thanks for the videos and screenshots shared. Excellent!

I'm a bit late to join the conversation, so a few quick ones I've seen discussed in #44420, just to chime in on:

* I think it's good to start with the eye. Even if it isn't entirely clear, that fact itself helps balance the prominence of the feature and lets us get this out there as a first version. As Jay alludes to, the site editor could get an I/A restructuring in the future, at which point we can revisit where and how the style book looks. 
* I would echo that we should consider using the dark gray background of focus mode, rather than the light gray background used here. This is mainly to avoid introducing new design elements. A different way of saying it is, if we should use a light background here, we should also use it for "focus mode", and "zoom out" modes. Conceptually, these are all related as they shift the main contents around.
* One thing I liked about #44420 was that it didn't include a "back" button. The main challenge with the back button is that it implies history/browser navigation, where as this is more of a "mode" that you can enter. In that light, the modal-esque close button made more sense. 

This is a riff on that and this PR, _very_ quick and dirty:

<img width="1567" alt="Screenshot 2022-11-28 at 15 52 46" src="https://user-images.githubusercontent.com/1204802/204308289-b0282fea-83a3-456f-abff-5adcd2e47a01.png">

I included in that a "welcome screen", which is mainly a way of saying: we could make that card itself be sort of interactive. Not only does it contain the tabs and the close button, but it contains info on how to use it. Not a strong opinion, and not a blocker for this PR — the main take-aways are the dark background.

Thank you Javi and Robert!

noisysocks commented:
Awesome, thanks for giving it a spin and for the feedback! I'll get this ready to land as a first version. Hopefully in time for SotW.

noisysocks committed:
Add Style Book

noisysocks committed:
Disable toolbar butons when Style Book is open

javierarce commented:
I've updated the Figma file with @jasmussen's [iteration](https://www.figma.com/file/3CtpCjdXvt1ZQLu7P1m13k/Global-styles-summary-view-for-blocks-and-styles?node-id=886%3A16948), including a proposal for how to show the block types when the canvas is resized.

![image](https://user-images.githubusercontent.com/4933/204570194-a7611138-96f9-4d41-8caf-b871b52670de.png)

noisysocks committed:
Disable Style Book when in Code Editor mode

noisysocks committed:
Make cursor a hand when hovering an example

noisysocks committed:
Remove top and bottom margin from block inside preview

noisysocks committed:
Add border beneath tab bar, make tab bar scroll on mobile

noisysocks committed:
Improve performance

noisysocks committed:
Improve a11y

noisysocks committed:
Add E2E tests

noisysocks commented:
Ready for review! I updated the PR description and included a re-recorded video.

> * If you click several blocks in the canvas and then use the Global Styles sidebar back button to reach the top level (or the list of blocks), you'll be presented with every single block settings panel you clicked before. I think we shouldn't stack those clicks.

Totally agree with this. It should be doable but requires a few tricky changes to how `Navigator` works in `@wordpress/components`. Let's tackle it in a follow-up PR.

> * The content within each section should align vertically with the label:

It's difficult to align the labels _perfectly_ because blocks can have different margins, line heights, etc. I implemented an approach which unsets the top and bottom margin of the block. I think this works well enough.

> I included in that a "welcome screen", which is mainly a way of saying: we could make that card itself be sort of interactive. Not only does it contain the tabs and the close button, but it contains info on how to use it. Not a strong opinion, and not a blocker for this PR — the main take-aways are the dark background.

I didn't implement this as I have a few concerns 😀

It's strange to me that it has Gutenberg branding. This feature will ultimately ship in WordPress as a part of the Site Editor. I imagine many will see it and think "what's Gutenberg?"

I also don't love that it adds an extra step between opening the Style Book and previewing the styles. Could we instead have onboarding text within the Style Book somewhere?

At any rate, let's do this in a follow-up PR.

> including a proposal for how to show the block types when the canvas is resized.

I skipped the part of this where the tabs turn into a dropdown. This behaviour should be something that `TabPanel` does automatically, so let's implement it in a follow-up `@wordpress/components` PR. For now as an interim solution I just made the tabs have `overflow: auto`.

jasmussen commented:
Having slept upon it, I was reminded that Gutenberg is just the plugin name, so despite my previous doodle, we should probably omit the logo and "Gutenberg" moniker after all. Sorry about that detour 🙈

javierarce commented:
> Having slept upon it, I was reminded that Gutenberg is just the plugin name, so despite my previous doodle, we should probably omit the logo and "Gutenberg" moniker after all. Sorry about that detour 🙈

No problem, I'll try to come up with an illustration to replace the logo.

jameskoster commented:
The tabs look a bit strange on a theme with a dark background, and the close button becomes a bit hidden:

<img width="1145" alt="Screenshot 2022-11-30 at 12 19 37" src="https://user-images.githubusercontent.com/846565/204794695-a5cff278-9e22-4248-bd71-71ea81e0b948.png">


Should the tab bar have a hard-coded white background?

---

Instead of disabling the Inserter button and other tools when the Style Book is open, should we just hide them?

| Disabled | Hidden |
| --- | --- |
| <img width="1145" alt="Screenshot 2022-11-30 at 12 22 40" src="https://user-images.githubusercontent.com/846565/204795418-c84aa42b-4503-4385-9202-1f98fdb913e3.png"> | <img width="1136" alt="Screenshot 2022-11-30 at 12 23 10" src="https://user-images.githubusercontent.com/846565/204795458-f176ac1d-e50c-4a69-a9d4-0a9c8120806b.png"> |

---

A small detail: Is there anything we can try to make it more obvious that the block names are part of the UI rather than the block itself? They can be a bit too similar when the theme uses a sans-serif font. All-caps? Monospace font? Separator?

| All-caps | Monospace | Separator |
| --- | --- | --- |
| <img width="552" alt="Screenshot 2022-11-30 at 12 29 07" src="https://user-images.githubusercontent.com/846565/204797144-fa3e1223-f66f-4b8a-9e22-d8ebf39d3da8.png"> | <img width="363" alt="Screenshot 2022-11-30 at 12 30 02" src="https://user-images.githubusercontent.com/846565/204797147-021b698c-4c07-4510-b180-7e23251e7627.png"> | <img width="418" alt="Screenshot 2022-11-30 at 12 31 29" src="https://user-images.githubusercontent.com/846565/204797151-dc7b2ab8-ea91-4c7e-935a-a6f7b8f20517.png"> |

---

I also must say that the eye icon really doesn't do it for me. I associate it with visibility settings and don't see any relationship between it and the Style Book feature. Is there another option to try?

---

Edit: One more note – appreciate it's follow-up territory – but it would be really nice if you could search for blocks by name.

jasmussen commented:
> No problem, I'll try to come up with an illustration to replace the logo.

I think Jay mentioned separately, that it may just be best to start without any welcome screen. Sorry for that detour.

javierarce commented:
> I also must say that the eye icon really doesn't do it for me. I associate it with visibility settings and don't see any relationship between it and the Style Book feature. Is there another option to try?


I don't like it either. What about introducing a book icon like this one?

![image](https://user-images.githubusercontent.com/4933/204896374-1d1504f6-7437-4350-bb98-48201f067674.png)

jasmussen commented:
I recall the eye having drawn controversy ever since we tried it as an icon-only preview button in very early drafts 😅 

But similarly, it's been used in the Global Styles Figma for years at this point, and I think we deserve to at least try it out in the plugin for a bit, get a feel for it. It's easy to revisit. What do you say?

jameskoster commented:
I think the book metaphor works much better 🤷‍♂️

javierarce commented:
> I think the book metaphor works much better 🤷‍♂️

I agree. The eye seems much more connected with the idea of a preview, like [in this issue](https://github.com/WordPress/gutenberg/issues/42919)

> Should the tab bar have a hard-coded white background?
+1 

> Instead of disabling the Inserter button and other tools when the Style Book is open, should we just hide them?

+1

> A small detail: Is there anything we can try to make it more obvious that the block names are part of the UI rather than the block itself? They can be a bit too similar when the theme uses a sans-serif font. All-caps? Monospace font? Separator?

All-caps looks nice. I think that, plus their arrangement in a row could be enough.

> One more note – appreciate it's follow-up territory – but it would be really nice if you could search for blocks by name.

I'll add a note to the Figma file to explore that idea.

noisysocks committed:
Make example titles uppercase

noisysocks committed:
Hide toolbar when Style Book is open

noisysocks committed:
Make tabs always have white background

noisysocks committed:
Update tests

noisysocks commented:
I've made the titles uppercase, made the editor toolbar disappear when the Style Book is open, and made the tab bar always have a white background.

Where can I find the book icon? Happy to update that, but couldn't find it in the Figma.

<img width="1258" alt="Screenshot 2022-12-02 at 12 15 38" src="https://user-images.githubusercontent.com/612155/205195307-2ec047a5-cc8b-4e3f-a6d2-fd54bb6e953b.png">

andrewserong commented on packages/edit-site/src/components/block-editor/index.js:
```diff
@@ -211,54 +230,71 @@ export default function BlockEditor( { setIsInserterOpen } ) {
-					</__unstableBlockNameContext.Consumer>
-				</__unstableBlockToolbarLastItem>
-			</BlockTools>
+			<StyleBook.Slot>
```
Will there be other use cases for taking over the editor canvas in the future? If so, would it be worth giving the slot a more generic name, since it's the parent of both the style book and the site editor canvas... perhaps something like `EditorCanvas.Slot`?

noisysocks replied:
That's a good question!

I'm using the presence of fills in the slot to determine if the Style Book is open or not (`useHasStyleBook`). This lets us hide controls in the editor toolbar and display "Style Book" in the centre. If we go with a more generic slot e.g. `EditorCanvas` we'd need to come up with an alternative approach for that. Potentially you could have an additional slot in the toolbar e.g. `EditorToolbar` and `StyleBook` fills both slots.

I can't think of any immediate future need for this though. Since this slot isn't exposed as a public API outside of `@wordpress/edit-site` I'm inclined to do the most minimal thing possible and implement a more generic API if/when the need arises.

Do you think it's worth adding an inline comment that encourages a future developer working in this area to not be afraid of a change like this? e.g. `// Potentially this slot could be more generic e.g. EditorCanvas.Slot`.

andrewserong replied:
> I can't think of any immediate future need for this though. Since this slot isn't exposed as a public API outside of @wordpress/edit-site I'm inclined to do the most minimal thing possible and implement a more generic API if/when the need arises.

Agreed! It's highly likely that it isn't something we'll need any time soon.

> Do you think it's worth adding an inline comment that encourages a future developer working in this area to not be afraid of a change like this? e.g. // Potentially this slot could be more generic e.g. EditorCanvas.Slot.

Yes, definitely not a blocker, but I think adding a comment like that would be good — I know I find those sorts of comments super helpful when I'm working out which things are okay for me to hack around with 😀

andrewserong reviewed (commented):
Great work here @noisysocks, it's testing really well for me, and the code looks pretty solid! 👍

Just left a question about the naming of the slot, but it's not at all a blocker as I think it'd probably be fairly straightforward to change in a follow-up if need be.

noisysocks committed:
Add a little note of encouragement for future developers

jasmussen commented:
This is really solid. Here's a GIF showing the feature:

![status](https://user-images.githubusercontent.com/1204802/205250141-f7a9cb2e-5662-4d95-bbd9-85e9ef2096c6.gif)

A few small details. Can you add a `border-radius: $radius-block-ui;` to the frame?

<img width="147" alt="Screenshot 2022-12-02 at 09 30 35" src="https://user-images.githubusercontent.com/1204802/205250265-8370c84e-42c9-4d16-a459-a97fb7342414.png">

I wonder if this button should be a toggle, it feels pretty heavy. Not a strong opinion. And separately — not for this PR, just a note for design efforts in #43423 — we should refactor so those buttons (eye, ellipsis, X) have a 24x24px or 32x32px footprint. 

<img width="147" alt="Screenshot 2022-12-02 at 09 30 53" src="https://user-images.githubusercontent.com/1204802/205250283-207e4a25-26a9-447e-9f15-a9cc97bbf84d.png">

The x is slightly vertically misaligned here:

<img width="127" alt="Screenshot 2022-12-02 at 09 31 10" src="https://user-images.githubusercontent.com/1204802/205250646-6e192bc9-2459-422c-9984-93bd01d27c27.png">

Should we do without the hover style? The thin gray border here:

<img width="855" alt="Screenshot 2022-12-02 at 09 31 15" src="https://user-images.githubusercontent.com/1204802/205250701-d1a705a9-474a-40ea-80a0-182884bf901a.png">

If it's necessary, we should look into using some opacity magic so it blends better on colored backgrounds. 

This heading:
<img width="131" alt="Screenshot 2022-12-02 at 09 31 29" src="https://user-images.githubusercontent.com/1204802/205250859-5ce6af4f-4d16-4015-967a-deff64599806.png">

... should it be the same visual treatment as our labels? These:

<img width="267" alt="Screenshot 2022-12-02 at 09 36 21" src="https://user-images.githubusercontent.com/1204802/205250968-505cf69e-44fb-4d14-aa62-3fee02a07191.png">

They should probably also be `currentColor` so they inherit the font color and work in any background contexts.

Really nice work, these are small details that we can quickly address and ship. Thank you.

noisysocks committed:
Add border radius to Story Book

noisysocks committed:
Fix vertical alignment of close button

noisysocks committed:
Reduce font size and increase font weight of example labels

noisysocks committed:
Remove hover outline

noisysocks committed:
Use theme color for text instead of white or black

noisysocks commented:
Thanks Joen! I added border radius, fixed the close button alignment, removed the hover styling, and made the labels use the theme's text color. I made the Style Book labels use the same font weight as the sidebar labels (500) but used 13px instead of 11px as a font size because 11px felt wrong to me. We can iterate on the toggle buttons and hover effect—it's easy to change.

Here's the current state:

<img width="1258" alt="Screenshot 2022-12-05 at 11 29 10" src="https://user-images.githubusercontent.com/612155/205525388-ddd2b9db-9202-4f12-ac0c-ec4761b6d3f5.png">

andrewserong commented on packages/edit-site/src/components/style-book/style.scss:
```diff
@@ -0,0 +1,78 @@
+.edit-site-style-book__close-button {
+	position: absolute;
+	right: $grid-unit-10;
+	top: ($grid-unit-60 - $button-size) / 2; // ( tab height - button size ) / 2
```
Looks like this is generating a Sass deprecation warning for me locally. Would it be worth using a `calc()` rule here?

<img width="1067" alt="image" src="https://user-images.githubusercontent.com/14988353/205525855-32bfece2-d6e0-40fc-9259-d88202511f4a.png">

noisysocks replied:
Oh! Didn't know this was deprecated. Looks like `math.div` is appropriate here.

https://sass-lang.com/documentation/breaking-changes/slash-div

andrewserong replied:
Nice one, TIL about `math.div` 😄

noisysocks replied:
Look at us learning stuff.

ramonjd replied:
![image](https://user-images.githubusercontent.com/6458278/205526621-5daaa14f-e986-48c5-af3a-6eab55f663c6.png)

noisysocks committed:
Use math.div()

andrewserong reviewed (approved):
Code's looking good to me, and this is still testing nicely after the latest design refinements 👍

<img width="1278" alt="image" src="https://user-images.githubusercontent.com/14988353/205526370-1de1efbf-1dfc-48a3-a89e-76b14e8521c8.png">

I think this is a terrific feature, it's going to be so useful when making changes to block styles. Great work! ✨

ramonjd reviewed (approved):
Good stuff @noisysocks ! 

Things are working well and I can see the updates to my global styles changes in real time 👍 

The only comment I'd make for a follow up is to somehow inform users what the feature is and why it's there in the editor somewhere. Even a five-word explainer:

<img width="911" alt="Screen Shot 2022-12-05 at 11 54 30 am" src="https://user-images.githubusercontent.com/6458278/205527031-b1cda40e-8c3b-4f68-9e14-1524178b51d4.png">

For me, it's a bit of a "mystery button" otherwise, and the switch to demo content might throw some folks.

"Style book" is a terrific name, but doesn't, in itself, communicate much.

noisysocks commented:
> The only comment I'd make for a follow up is to somehow inform users what the feature is and why it's there in the editor somewhere. Even a five-word explainer:

Yeah I like the idea of the welcome screen in https://github.com/WordPress/gutenberg/pull/45960#issuecomment-1329244568 but would probably prefer to dial it back a bit so that there's no extra step. For example have some copy at the top of the Style Book that explains things.

jasmussen commented:
Thanks for the effort!

> but used 13px instead of 11px as a font size because 11px felt wrong to me. We can iterate on the toggle buttons and hover effect—it's easy to change.

Using 11px was not so much about the font size, but rather to not diverge from existing code. I understand there's no CSS class for the new label style, it's currently a CSS-in-JS `styledLabel` as I recall. But ideally it would all be the same in the name of consistency, so it could be a good followup.

jameskoster commented:
Agreed. The full appearance (size included) is what helps distinguish the labels as part of the UI.

bph commented:
The ‘story book’ is such a cool feature. And naming things is hard. 

The Storybook that has been in the block editor until now is the [interactive component tool for developers.](https://wordpress.github.io/gutenberg/?path=/docs/docs-introduction--page).
Could we possibly find a different name for the story book in his PR? Like “Style Guide”, “Block design guide”?

noisysocks commented:
Open to suggestions 😀 it's not too late to rename. FWIW, the implemented name is _Style Book_, not _Story Book_.

noisysocks commented:
@jasmussen @jameskoster I updated the font size in https://github.com/WordPress/gutenberg/pull/46341.

bph commented:
Ah Style book. My brain just wasn't slowing down long enough to even see the difference. Sorry about that.

madhusudhand commented:
This is amazing. Created #46393 and #46394 for the followup.

sunyatasattva commented:
Hey @noisysocks, sorry for intruding into the thread, but is there any way to disable Style Book for certain blocks? We can remove the `example` prop from the `block.json`, but that would make the live preview disappear from the inserter.

Some of our blocks are not style-able, or not style-able *yet*, so it'd be nice if there was an opt-out mechanism.

noisysocks commented:
The only way to opt out right now is by removing the `example` attribute. We can consider adding a different opt out mechanism. Just so that I understand the problem fully, what's your reason for not wanting a block to appear in Style Book?

sunyatasattva commented:
The main reason is that Global Styles are not quite supported, or are currently buggy, so we'd rather hide it temporarily, then give the users a false impression that they can use Global Styles on those blocks.

The secondary reason is that some of the blocks don't have any styles options whatsoever.

noisysocks commented:
Thinking about it some more I think an opt out mechanism might be inappropriate.

The point of Style Book is to see how style changes affect everything that might be inserted. It's equivalent to inserting all of the blocks in one's library into a page to test them. Even if a block doesn't adopt any global style settings, element styles may indirectly affect the block and users / theme developers should be aware of that impact.

bph commented:
Flagged for 'Single Post Dev Note`. It's a major enhancement for developers (their blocks show up into the style book) and the mechanism to opt-out via removing the 'example' attribute.

andrewserong commented:
## Dev note for 6.2

### Style Book preview of blocks in global styles

In WordPress 6.2 a new feature has been added to the global styles interface that allows a user to preview every block that can be inserted into their site. The Style Book can be accessed via the eye icon in the Styles header within global styles, and provides an at-a-glance preview of how global styles will affect the display of any block, without the user having to insert those blocks into a template.

Following the structure of the block inserter, the Style Book will be populated by every block that can be inserted via the block inserter, and that also an `example` set. The categories/tabs in the Style Book also follow the same categories used by the block inserter. For block developers that already set an `example` for custom blocks, no additional parameters are required for the block to be visible in the Style Book.

Within the Style Book, a user can click on a preview of a block and it will navigate them to the block-level global styles where the block's styling can be adjusted. Overall, this makes it easier for a user to preview changes to block-level global styles, without having to insert the block within the current template.

Even if a block has no direct styling controls, it is still valuable for blocks to appear in the Style Book, so that the user can see how the overall site design and styling settings will affect every block that can be inserted. For example, site-wide background and text colors, or font family or font sizes generally affect how all blocks will be displayed.

However, there still might be cases where block developers wish to hide blocks from being inserted and therefore also hide them from the Style Book. The two ways to hide blocks from the Style Book are:

* If the block has `supports.inserter`  set to `false` within the `block.json` file, then the block will be hidden from both the block inserter and the Style Book.
* If the block does not have an `example` set.

To read more on adjusting the `supports.inserter` value, see [the entry on block supports](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/#inserter) in the developer handbook.

The conditions of when to show or hide blocks from the Style Book are intentionally tied to the same logic as used by the block inserter. The principle behind the Style Book is that if a user can insert or view a preview of the block, then it should also be previewed within the context of global styles.

---

@bph here's a draft above — does that about cover it? Happy to continue to edit / add or remove things from this one as needed.

bph commented:
@andrewserong Thank you for drafting the dev note. I think it covers it. Maybe add an example block.json section 
so we wouldn't need to link out to the documentation for this single snippet. 
```
supports: {
    // Hide this block from the inserter.
    inserter: false
}
```
If you agree that this is suitable for a single dev note post, would you please create a draft on the Make Blog so we can get into the review process?

andrewserong commented:
Yes, of course, thanks @bph — I've added a draft for the Style Book dev note over in: https://make.wordpress.org/core/?p=102735&preview=1&_ppp=03ed6e05f5 (also moved the links to be inline and added the `block.json` example)

noisysocks commented:
Thanks for writing the note @andrewserong!
